### DISCUSSION TITLE
Proposal: Explicit RAII

### DISCUSSION THREAD

**Discussion by danmitreanu** [dotnet/csharplang#7321] [Language Ideas]
# RAII in C#

## Problem

RAII can currently be "emulated" in C# via the Dispose pattern, where you can define the initialization behavior in the constructor of a class and the tear-down (or "destructor"?) in the `Dispose()` method.

```csharp
class FileLock : IDisposable
{
    public FileLock() => File.WriteAllText("file.lock", string.Empty);
    public void Dispose() => File.Delete("file.lock");
}

void SomeFunc()
{
    using FileLock flock = new();
    // do stuff that require a lock
} // <- lock released here
```

There was [a proposal](https://github.com/dotnet/csharplang/discussions/1110) somewhere in 2017 that suggested adding a destructor to `ref struct`s. Ref structs are great for RAII considering they are also trapped on the stack. Allocating on the heap for RAII is useless extra cost. Ref structs can also be `IDisposable` since C# 8.0 (or somewhere around that time), which abolishes that cost.

The main issue with implementing a RAII-enabled type, or worse, a RAII-only type, is making sure the user of said type calls `Dispose()` or instantiates the type with a `using` statement. This prompts [worrying but valid hacks](https://stackoverflow.com/questions/20358401/throwing-exception-in-finalizer-to-enforce-dispose-calls).

## Proposed solution

Provide developers the means to create structures with deterministic destruction. Considering the nature of C#, they would of course come with some caveats, most of which would probably be shared with the limitations of `ref struct`s:

- A ref struct can't be the element type of an array.
- A ref struct can't be a declared type of a field of a class or a non-ref struct.
- ~A ref struct can't implement interfaces.~
- A ref struct can't be boxed to [System.ValueType](https://learn.microsoft.com/en-us/dotnet/api/system.valuetype) or [System.Object](https://learn.microsoft.com/en-us/dotnet/api/system.object).
- A ref struct can't be a type argument.
- A ref struct variable can't be captured by a [lambda expression](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/lambda-expressions) or a [local function](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/local-functions).
- ~A ref struct variable can't be used in an [async](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/async) method. However, you can use ref struct variables in synchronous methods, for example, in methods that return [Task](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.task) or [Task<TResult>](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.task-1).~ *TBD?*
- A ref struct variable can't be used in [iterators](https://learn.microsoft.com/en-us/dotnet/csharp/iterators).

The difference between this proposal and other similar proposals is the suggestion of a simple, elegant and non-intrusive solution: the `local` keyword. The `local` keyword would allow creation of RAII-enabled types:

```csharp
public local class FileLock { ... }
public local ref struct FileLockOnTheStack { ... }
```

The initialization would be best kept in the constructor, as it is intuitive. For the *destructor* (or *cleanup*?), there are a number of solutions:

1. As with `IDisposable`, provide a new interface that is used by the runtime to perform cleanup. Example: `public local class FileLock : ILocal`, `interface ILocal { void CleanUp(); }`. The runtime would call the method(s) in this interface once the instance goes out of scope. **Additional idea**: since the `local` keyword is already there, maybe make `local` classes and structs derive from `ILocal` by default?
2. Use the already available `IDisposable`, with the runtime calling `Dispose()` once the instance goes out of scope, basically like it currently works (read on...). **Additional idea**: same as above.
3. Use the old-fashioned, severely underused and mostly purposeless finalizer, `~FileLock()`.

Going back to point no. 2, if using the already available `IDisposable` is a possibility, why add this `local` stuff? Short answer: to explicitly specify that this structure WILL be deterministically destroyed when going out of scope.

The `local` keyword in the structure definition should inform the compiler that this type should be deterministically destructed.
Additionally, the `local` keyword should be enforced also on instantiation of local types to also make things clear for the code reader:

```csharp
void DoStuff()
{
    local FileLock flock = new();
    // ^ we now know this will be destroyed at the end of the method
    // stuff
}
```

***TLDR***: Think of `IDisposable` and how those types can be created even outside of `using` blocks, creating situations where they are not being disposed immediately by the runtime nor deterministically by the developer. Enable the creation of `local` types that remove any kind of doubt about when they are destroyed and allow the developer that implements them to make sure they are disposed when they should be.

**Comment by CyrusNajmabadi** [comment]
> The main issue with implementing a RAII-enabled type, or worse, a RAII-only type, is making sure the user of said type calls Dispose() or instantiates the type with a using statement. 

A way around thsi would be to just write an analyzer for your particular types that then enforce a proper coding pattern against them.  Now you would have the benefit of your code having these rules enforced, so you would not have any mistakes *and* the code would be clear as to its intent.

> The local keyword in the structure definition should inform the compiler that this type should be deterministically destructed.
Additionally, the local keyword should be enforced also on instantiation of local types to also make things clear for the code reader:

This is just a synonym for `using` at that point.  However, with the caveat that it sounds like you would expect the compiler to error if you forgot to state that at the use-site.  However, if that's what you would like thei compiler to do, this could be achieved today by putting a special `[LocalDisposable]` attribute on these types you care about (or annotated on the members that return them), and then having an analyzer that checks the use site to make sure it is actually cleaned right there :)

**Reply by danmitreanu** [reply]
Yep, that's certainly all doable. There's all sorts of ways to enforce this by playing around with Roslyn. Just like applying attributes on a property's underlying field was doable before the `[field: MyAttr]` thing by manually creating the backing field.
This is not really a matter of whether it's doable or not, I'm sure it is with, especially when having the .NET Compiler Platform SDK at your disposal. It is a proposal for a feature that might make the lives of some developers easier with some out-of-the-box functionality. This is of course a matter of personal opinion and so it is very much debatable.

**Reply by CyrusNajmabadi** [reply]
> It is a proposal for a feature that might make the lives of some developers easier with some out-of-the-box functionality

So, IMO, attributes/analyzers *are* out of the box :)   

And they package up *super nicely* in simple packages that you, your team, or anyone else can reference if they find value in it.  It also means that if the 1st party 'out of hte box' solution isn't a fit for you (for example, you need to tweak certain cases where it fires) that is fully at your disposal.

Importantly, when a feature is effectively "work like C#, but with *more* restrictions that act as guardrails", that's *really* the ideal place to just use an analyzer.  

1. it requires no change to the language (incredibly costly, and hard to justify depending on how niche the scenario).
2. it works downlevel for like 10+ years worth of compilers.
3. it works without rev'ing the language version.  So someone can use this, even if they're locked to C#8.
4. it's very easy to distribute.
5. as mentioned above, it is much easier to make things flexible to support a variety of cases.  This is something that is effectively not possible as a language feature.  For example, you could have booleans with your attribute that slightly configure behavior/policy in different circumstances.
6. it can usually be done by an experienced user in an hour or so, and a novice user in about a day.  That's orders of magnitude less effort, and enables extremely fast turnaround on making the feature and trying it out.

I def recommend trying it out!  :)

Once you get comfortable with teh API, you'll find that you can scratch these itches easily and get the benefits from them immediately (instead of potentially having to wait 5+ years).

**Comment by deerchao** [comment]
It would be very nice to be able to write this if we can automatically destruct those resource objects:

```csharp
SomeMethod(new SomeResource1(), new SomeResource2(), new SomeResource3(), new SomeResource4());
// even better if you have factory methods to create those resources
SomeMethod(R1(), R2(), R3(), R4());
```

instead of:

```csharp
using var r1 = new SomeResource1();
using var r2 = new SomeResource2();
using var r3 = new SomeResource3();
using var r4 = new SomeResource4();
SomeMethod(r1, r2, r3, r4);
```
or

```csharp
using var r1 = R1();
using var r2 = R2();
using var r3 = R3();
using var r4 = R4();
SomeMethod(r1, r2, r3, r4);
```

**Reply by huoyaoyuan** [reply]
This is a strongly breaking change that would never happen. There are cases that need to avoid `Dispose` explicitly.

**Reply by deerchao** [reply]
We can explicitly mark the class or object as RAII with `local` keyword to an `IDisposable`, just like @danmitreanu proposed.

**Comment by Richiban** [comment]
Another solution to this (and something that I've long wished C# had) is some kind of an equivalent to F#'s [InlineIfLambda](https://learn.microsoft.com/en-us/dotnet/fsharp/language-reference/functions/inline-functions#inlineiflambda).

Basically something that means that you could implement your resource usage like this:

```cs
class FileLock
{
    public static void WithLock([Inline] Action action)
    { 
        File.WriteAllText("file.lock", string.Empty);
        action();
        File.Delete("file.lock");
    }
}

void SomeFunc()
{
    FileLock.WithLock(() => {
        // do stuff that require a lock
    }); // <- lock released here    
}
```

and it would be zero-overhead because it just gets rewritten to 

```cs

void SomeFunc()
{
    File.WriteAllText("file.lock", string.Empty);
    // do stuff that require a lock
    File.Delete("file.lock");
}
```

**Reply by CyrusNajmabadi** [reply]
If you do:

```c#
class FileLock
{
    public static void WithLock<TArgs>([Action<TArgs> action, TArgs args)
    { 
        File.WriteAllText("file.lock", string.Empty);
        action(args);
        File.Delete("file.lock");
    }
}

void SomeFunc()
{
    FileLock.WithLock(static args => {
        // do stuff that require a lock, and args.a, args.b, args.c
    }, args: (a, b, c)); // <- lock released here    
}
```

Then you will get affectively that.  First, the lambda is static, so only incurs an allocation once,  no matter how many times this is called.  The use of struct-args means you get custom specialized emit for all different callsites (just like as if it had  been inlined).  And thanks to modern jit optimizations, it can 'see through' this call, and effectively inline if it feels it is worthwhile.

**Reply by TahirAhmadov** [reply]
https://github.com/dotnet/csharplang/discussions/4187
https://github.com/dotnet/csharplang/discussions/6287

**Reply by timcassell** [reply]
Also #8816

**Reply by TahirAhmadov** [reply]
There's another avenue, around adding exceptions to `finally` blocks (optionally), so that a similar behavior can be achieved with `using(var scope=BeginScope())` pattern. I found these related discussions:
https://github.com/dotnet/runtime/issues/75729
https://github.com/dotnet/roslyn/issues/4890
https://github.com/dotnet/csharplang/discussions/213
https://github.com/dotnet/csharplang/discussions/216
https://github.com/dotnet/csharplang/discussions/434
I still hope this gets added at some point, because it gets rid of a lot of boilerplate and reduces the number of method stack frames.

**Reply by deerchao** [reply]
Consider if you need more than 1 locks, like 4 in my example, it effectively turns into callback hell.

**Comment by danmitreanu** [comment]
Update 2 years later: there are a number of features that have been added to the language that could benefit from this, for example `Lock.EnterScope` (which has a usage pattern very similar to the C++ RAII-based `std::unique_lock`, for example).
